File: bootstrap-maven-spring-cucumber-junit5/src/test/java/cucumber/stepdefs/StepDefsContext.java
package cucumber.stepdefs;

import java.util.HashMap;
import java.util.Map;

public enum StepDefsContext {

    CONTEXT;

    private final ThreadLocal<Map<String, Object>> givenObjects = ThreadLocal.withInitial(HashMap::new);
    private final ThreadLocal<Map<String, Object>> thenObjects = ThreadLocal.withInitial(HashMap::new);

    public <T> T givenObject(Class<T> clazz) {
        return clazz.cast(givenObjects.get()
                .get(clazz.getName()));
    }

    public <T> void givenObject(T object) {
        givenObjects.get()
                .put(object.getClass().getName(), object);
    }

    public <T> T result(Class<T> clazz) {
        return clazz.cast(thenObjects.get()
                .get(clazz.getName()));
    }

    public <T> void result(T object) {
        thenObjects.get()
                .put(object.getClass().getName(), object);
    }

    public Throwable throwable() {
        return Throwable.class.cast(thenObjects.get()
                .get(Throwable.class.getName()));
    }

    public void throwable(Throwable throwable) {
        thenObjects.get()
                .put(Throwable.class.getName(), throwable);
    }

    public void reset() {
        givenObjects.get()
                .clear();
        thenObjects.get()
                .clear();
    }
}

File: bootstrap-gradle-spring-cucumber-kotlin-junit5/src/test/kotlin/cucumber/stepdefs/StepDefsContext.kt
package cucumber

import kotlin.reflect.KClass

enum class StepDefsContext {
    CONTEXT;

    private val givenObjects = ThreadLocal.withInitial<MutableMap<String, Any>> { HashMap() }
    private val thenObjects = ThreadLocal.withInitial<MutableMap<String, Any>> { HashMap() }

    fun <T : Any> givenObject(clazz: KClass<T>): Any? {
        return givenObjects.get()[clazz.toString()]
    }

    fun <T : Any> givenObject(obj: Any) {
        givenObjects.get()[obj.javaClass.toString()] = obj
    }

    fun <T : Any> result(KClass: Class<T>): Any? {
        return thenObjects.get()[KClass.name]
    }

    fun <T : Any> result(obj: T) {
        thenObjects.get()[obj.javaClass.toString()] = obj
    }

    fun throwable(): Throwable {
        return thenObjects.get()[Throwable::class.simpleName] as Throwable
    }

    fun throwable(throwable: Throwable) {
        thenObjects.get()[throwable.javaClass.toString()] = throwable
    }

    fun reset() {
        givenObjects.get()
                .clear()
        thenObjects.get()
                .clear()
    }
}
